doc/developer/design/20220226_multiwriter_persist.md

```diff
@@ -0,0 +1,205 @@
+
+* `CreateCollection(id)` -> ReadHandle: Creates a new persistent collection
+  identified by `id` and returns a read handle. After a persistent collection
+  has been created anyone can publish new batches into its pending area.
+* `AcquireReadHandle(id) -> ReadHandle`: provides a read handle at its current
+  `since` frontier and referring at the latest merged area version.
+* `DeleteCollection(id)`: Deletes a persistent collection identified by `id`
+
+### Reading
+
+A read handle of a persistent collection is described by a merged area version
+and a compaction frontier. **Readers of a collection must use the batches
+referred to by their current merged area version for any reads at times not
+beyond the merged area's `upper`. For times beyond `upper` the pending area is
+used instead.**
+
+This behavior is the key to allow the PERSIST CONTROLLER to decide with
+certainty which batches are reachable and perform background garbage collection
+of unreachable batches.
+
+When a reader wants to read at a time `t` that is beyond the merged area's
+`upper` there might be multiple batches in the pending area that include that
+time. This can happen in the case of multiple writers writing out a definite
+collection. Readers are free to choose which batches to select based on any
+strategy. For example, a reader could pick a writer shard to follow and only
+switch to a different shard if its current shard is falling behind too much.
+The trade-off for switching which shard is followed is the additional work that
+needs to be done if the frontiers of the batches don't exactly line up.
+
+Readers continually try to discover newly minted merged area versions and
+increase their associated version to that, indicating that they will never use
+the batch metadata of any previous version.
+
+Readers are also able to downgrade their compaction frontier indicating that
+they no longer care to distinguish updates that happened not beyond the
+frontier.
+
+Both compaction frontier downgrades and merged area version downgrades are
+durably persisted.
+
+A read handle can be created by cloning an existing read handle. The newly
+created read handle will have its own independent merged area version and
+compaction frontier with their initial values set to the values of the read
+handle it was cloned from. This is also a durable operation.
+
+### Writing
+
+A writer to a persistent collection can initialize itself completely
+independently simply by knowing the `id` of the collection. Writers are
+identified by a name and it is up to the user to ensure that unique names are
+chosen.
+
+A writer writes to the collection by simply publishing new batches into its
+corresponding pending area shard.
+
+## Background maintenance
+
+For each persistent collection an asynchronous background compaction and
+garbage collection process is run by the PERSIST CONTROLLER. This controller is
+a singleton process and it must be ensured that only one runs at a given time.
```

Comment:
Can we get around this restriction by storing the version metadata in a proper database with ConditionalPut semantics? If you go to update the version metadata and someone else beats you to it, then you know that you’re no longer responsible for compaction and you commit suicide.

Reply:
Yeah that would work, essentially making the database be the singleton process (or pretend to be if we're talking about a distributed database).